### PR TITLE
ENH/REF/BUG Discrete models llnull improvements

### LINKED ISSUE #3476: BUG: convergence failure with llnull in Poisson
see mailing list Poisson example
We can work around convergence failure with default settings in the main model, but we cannot influence the fit of the null model if that doesn't converge.

https://groups.google.com/d/msg/pystatsmodels/FAnjBe_RynE/e8hWinbZCQAJ

We can compute the constant directly from the data without optimization in cases like Poisson, and I guess most other GLM type models, maybe not when offset and exposure are used. And I'm not sure about noncanonical links (which might need a detour)

### LINKED ISSUE #3533: overflow encountered in exp [Poisson]
I have this error and my fit doesn't work. I don't know the reason (same problem in negativebinomial). Stats,model 0.8.0

    import pandas as pd
    features = ['pp']
    df = pd.read_csv('bug.csv')
    X = (df[features] - df[features].mean())/df[features].std()
    y = df['num'].values
    X_sm = sm.add_constant(X[features].copy())
    model = sm.Poisson(y, X_sm, ) #
    res = model.fit()

Thanks
[bug.csv.zip](https://github.com/statsmodels/statsmodels/files/816845/bug.csv.zip)

    /Library/Frameworks/Python.framework/Versions/3.4/lib/python3.4/site-packages/statsmodels/discrete/discrete_model.py:901: RuntimeWarning: overflow encountered in exp
    /  return stats.poisson.cdf(y, np.exp(X))
    //Library/Frameworks/Python.framework/Versions/3.4/lib/python3.4/site-packages/statsmodels/discrete/discrete_model.py:1175: RuntimeWarning: overflow encountered in exp
    /  L = np.exp(np.dot(X,params) + exposure + offset)
    //Library/Frameworks/Python.framework/Versions/3.4/lib/python3.4/site-packages/statsmodels/discrete/discrete_model.py:1112: RuntimeWarning: overflow encountered in exp
    /  L = np.exp(np.dot(X,params) + offset + exposure)
    //Library/Frameworks/Python.framework/Versions/3.4/lib/python3.4/site-packages/scipy/stats/_discrete_distns.py:451: RuntimeWarning: invalid value encountered in greater_equal
    /  return mu >= 0
    //Library/Frameworks/Python.framework/Versions/3.4/lib/python3.4/site-packages/statsmodels/base/optimizer.py:230: RuntimeWarning: invalid value encountered in greater
    /  oldparams) > tol)):
    //Library/Frameworks/Python.framework/Versions/3.4/lib/python3.4/site-packages/scipy/stats/_distn_infrastructure.py:875: RuntimeWarning: invalid value encountered in greater
    /  return (self.a < x) & (x < self.b)
    //Library/Frameworks/Python.framework/Versions/3.4/lib/python3.4/site-packages/scipy/stats/_distn_infrastructure.py:875: RuntimeWarning: invalid value encountered in less
    /  return (self.a < x) & (x < self.b)
    //Library/Frameworks/Python.framework/Versions/3.4/lib/python3.4/site-packages/scipy/stats/_distn_infrastructure.py:1814: RuntimeWarning: invalid value encountered in less_equal
    /  cond2 = cond0 & (x <= self.a)

### LINKED ISSUE #3919: BUG/Design: integer overflow when endog is int
In countmodels the endog is int32 in a simple example. We don't currently promote to float.

In GeneralizedPoisson, the hessian produces wrong results in some cases for dldada because of integer overflow.

Is there a problem if we always promote endog in count models to float64? or at least promote if int?

The problem shows up if there are some computation that only involve endog. e.g.
`y**2 * (y - 1) ` overflows, while `y**2. * (y - 1.)` looks fine

(also never trust an integer power)

(several hours of debugging to see what causes the "sometimes" bug)

### PR DESCRIPTION
**Pull request by josef-pkt**
This PR refactors llnull to use start_params provide by `model._get_start_params_null`.
The start_params for the null model can be obtained by moment estimators, which should make it more robust.
closes related issues: 
#3533 second part is for llnull, first part is convergence failure with defaults (where users can work around)
#3476 Poisson llnull, mailing list thread has also discussion on MNLogit
#3572 NB2 llnull

related issue #3751 (is more general, e.g. penalization, constraints. I don't know yet what should/needs to be done)

This also includes now refactoring and fixes that I ran into:

#3919 avoid integer overflow in GP hessian, REF set endog, exog dtypes to at least float64
#3918 start_params needed to to provide transformed params if transformparams is used internally

`closes #3533, closes #3476, closes#3572, closes #3919`   ("closes" is needed each time)

### CODE REVIEW

**Comment by josef-pkt** [comment]
Bonus: I'm writing moment estimators for the dispersion coefficient, NB1, NB2, NBP, GPP.
Those are currently designed for being hidden inside `_get_start_params_null`, but might be useful standalone (based on resid and mu).
similar to chisquare overdispersion scale in GLM, and score_tests for excess dispersion.

aside: count models don't have resid_pearson (discrete binary models have it)

**Comment by josef-pkt** [comment]
all red, problem in `statsmodels.discrete.tests.test_sandwich_cov` which I didn't run locally.
some failing llf comparisons and some shape mismatches.

my new tests (locally) still have convergence problems or create a lot of warnings, test case is definitely not nice for exp. (NB1 dispersion coefficient is above 2000)
One problem is convergence conditions are not very strict to test different start_params at high tolerance.
The second problem is that we get warnings in corner cases, that should be silenced if those are "expected".

**Comment by josef-pkt** [comment]
test error py 2.7: `AttributeError: 'list' object has no attribute 'copy'`
start_params is not converted to ndarray

**Comment by josef-pkt** [comment]
The last commit should finish with the main count models.

I'm not sure I will figure out the remaining null start_params for this PR, binary, MNLogit and ZeroInflated.

a possible additional test is to check that null start_params are close to fit params, which should be fine if one-step moment estimators are close to MLE.

another open issue is better preliminary fit to get the start_params for the final model.

In the unit tests in this PR, I avoided problems by hardcoding the start_params for the full model. The test classes could check generically for convergence in this case, future task when we have a better way to get default start_params. In my initial script for this PR I used `method='nm'` to get reasonable start_params that work with "bfgs" or "newton".

**Comment by josef-pkt** [comment]
2 widespread test failure
TestNegativeBinomialNBP2Null  nan in first llf null (w/ start_params_null)
TestGeneralizedPoissonNull  nan in second llf null  (comparison)

test pass in one python version on travis, and they pass locally for me

**Comment by josef-pkt** [comment]
on appveyor 2 out of 3 are green, one has the same two failures as travis

right now I don't see a reason why bfgs should converge in one case but not the other with similar starting values. From what I can see locally there shouldn't be a difference in the model or model parameters. 
It fails with scipy as recent as 0.17 on travis, the two passing on appveyor and the one passing on travis use scipy 0.19. I'm using scipy 0.18.1. So maybe the fixes in bfgs are not old enough to use it for this difficult test case.

needs more "unit" tests for intermediate results.

**Comment by josef-pkt** [comment]
related to this: I did not add an option to change the optimizer arguments in the llnull fit

Maybe I need to add it after all if start_params_null are not enough.

**Comment by josef-pkt** [comment]
New commit uses nm and newton instead of bfgs.

This is mainly to check the outcome across versions. I don't think it's a good default. "bfgs" looks better for recent enough scipy versions.

If it works, then we might have to add `set_null_optim_kwds` or `set_optim_kwds_null` to get around a bfgs default

**Comment by josef-pkt** [comment]
travis stopped on all versions
```
No output has been received in the last 10m0s, this potentially indicates a stalled build or something wrong with the build itself.
Check the details on how to adjust your build configuration on: https://docs.travis-ci.com/user/common-build-problems/#Build-times-out-because-no-output-was-received
The build has been terminated
```

maybe inf or nan in linalg when computing newton that hangs (although MKL usually raises instead of hangs)

**Comment by josef-pkt** [comment]
amended last commit and force pushed to switch back to 'bfgs', try different settings

**Comment by josef-pkt** [comment]
Another way to avoid spending (wasting) a lot of time fixing unit tests for extreme cases is to skip the two tests for scipy < 0.18. 
Anyone who wants good optimizers needs a recent scipy.

**Comment by josef-pkt** [comment]
everything green

However, I decided to add `set_null_options` to override everything in an official way.
`results.set_null_options(llnull=None, attach_results=True, **kwds)` 
this should have `_cache` manipulation to reset generated cached attributes.
`kwds` are fit keywords that override default settings
attach_results can set the internal `_attach_nullmodel` flag

no computation, computation are lazily triggered if llnull is used directly or indirectly

the intention is mainly to have an official override if the defaults fail, as in the issues that triggered this refactoring

**Comment by josef-pkt** [comment]
`set_null_option` added, plan is to merge tomorrow after an quick review

@denadai2 this should fix the llnull failures/nans in your issues, but I haven't tried with your dataset yet.
It's a bit later than we would hope for but it was a busy summer for me.

**Comment by josef-pkt** [comment]
The llnull problem with Poisson in #3533 is fixed by this, the new null start_params allow convergence and non-nan llnull.

(In my notebook for #3533 I had used scipy '0.17.0', now I'm using 0.18.1, which might also improve convergence.)

File: statsmodels/discrete/discrete_model.py [hunk 1]
```diff
@@ -3166,6 +3228,45 @@ def llr(self):
...
 
+    def set_null_options(self, llnull=None, attach_results=True, **kwds):
+        """set fit options for Null (constant-only) model
+
+        This resets the cache for related attributes which is potentially fragile.
+
+        Parameters
+        ----------
+        llnull : None or float
+            If llnull is not None, than the value will be directly assigned to
```
Comment (josef-pkt) [inline statsmodels/discrete/discrete_model.py]: typo than -> then

File: statsmodels/discrete/discrete_model.py [hunk 2]
```diff
@@ -3166,6 +3228,45 @@ def llr(self):
...
+            the cached attribute "llnull".
+        attach_results : Bool
+            Sets and internal flag whether the results instance of the null
```
Comment (josef-pkt) [inline statsmodels/discrete/discrete_model.py]: and -> an

File: statsmodels/discrete/discrete_model.py [hunk 3]
```diff
@@ -3166,6 +3228,45 @@ def llr(self):
...
+            model is attached. By default without calling this method, the
```
Comment (josef-pkt) [inline statsmodels/discrete/discrete_model.py]: model is attached -> model should be attached

**Comment by josef-pkt** [comment]
using a docstring template for internal functions is not so useful, because the docstring is not next to the source and the main audience are developers.
